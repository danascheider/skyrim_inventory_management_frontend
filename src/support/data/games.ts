import { ResponseGame as Game } from '../../types/apiData'

export const emptyGames: Game[] = []

export const allGames: Game[] = [
  {
    id: 32,
    user_id: 412,
    name: 'My Game 1',
    description: 'This is a game with a description',
    created_at: new Date(),
    updated_at: new Date(),
  },
  {
    id: 51,
    user_id: 412,
    name: 'My Game 2',
    description: null,
    created_at: new Date(),
    updated_at: new Date(),
  },
  {
    id: 77,
    user_id: 412,
    name: 'Game with a really really really really really long name',
    description:
      'Cum audissem Antiochum, Brute, ut solebam, cum M. Pisone in eo gymnasio, quod Ptolomaeum vocatur, unaque nobiscum Q. frater et T. Pomponius postmeridianam conficeremus in Academia',
    created_at: new Date(),
    updated_at: new Date(),
  },
]
