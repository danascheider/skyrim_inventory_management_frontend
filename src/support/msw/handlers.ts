import {
  postGamesSuccess,
  postGamesUnprocessable,
  postGamesServerError,
  getGamesEmptySuccess,
  getGamesAllSuccess,
  getGamesServerError,
  patchGameSuccess,
  patchGameUnprocessableEntity,
  patchGameNotFound,
  patchGameServerError,
  deleteGameSuccess,
  deleteGameNotFound,
  deleteGameServerError,
} from './games'
import {
  postShoppingLists,
  postShoppingListsUnprocessable,
  postShoppingListsServerError,
  getShoppingLists,
} from './shoppingLists'

export {
  postGamesSuccess,
  postGamesUnprocessable,
  postGamesServerError,
  getGamesEmptySuccess,
  getGamesAllSuccess,
  getGamesServerError,
  patchGameSuccess,
  patchGameUnprocessableEntity,
  patchGameNotFound,
  patchGameServerError,
  deleteGameSuccess,
  deleteGameNotFound,
  deleteGameServerError,
  postShoppingLists,
  postShoppingListsUnprocessable,
  postShoppingListsServerError,
  getShoppingLists,
}
