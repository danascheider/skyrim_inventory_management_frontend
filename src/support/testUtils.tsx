import { type ReactElement } from 'react'
import { BrowserRouter } from 'react-router-dom'
import { JSDOM } from 'jsdom'
import { render } from '@testing-library/react'
import { LoginContext } from '../contexts/loginContext'
import { User } from 'firebase/auth'
import testProfileImg from './testProfileImg.png'

export const BASE_APP_URI = 'http://localhost:5173'

declare global {
  namespace NodeJS {
    interface Global {
      document: Document
      window: Window
    }
  }
}

const setDom = () => {
  const dom = new JSDOM('<!doctype html><html><body></body></html>', {
    url: 'http://localhost:5173',
  })

  global.window = dom.window as unknown as Window & typeof globalThis
  global.document = dom.window.document
}

/*
 *
 * Global Test Data
 *
 */

export const requireLogin = () => {
  /* noop */
}

export const testUser = {
  uid: 'edna',
  displayName: 'Edna St. Vincent Millay',
  email: 'edna@gmail.com',
  photoURL: testProfileImg,
  emailVerified: true,
  getIdtoken: () =>
    new Promise<string>((resolve, _reject) => resolve('xxxxxxx')),
} as unknown as User

/*
 *
 * Test Renderers
 *
 */

const renderWithJSDOM = (ui: ReactElement) => {
  setDom()

  return render(ui)
}

export const renderWithRouter = (ui: ReactElement) =>
  renderWithJSDOM(<BrowserRouter>{ui}</BrowserRouter>)

export const renderAuthenticated = (ui: ReactElement, authLoading = false) =>
  renderWithRouter(
    <LoginContext.Provider
      value={{
        user: testUser,
        token: 'xxxxxxx',
        requireLogin,
        authLoading,
      }}
    >
      {ui}
    </LoginContext.Provider>
  )

export const renderUnauthenticated = (ui: ReactElement) =>
  renderWithRouter(
    <LoginContext.Provider
      value={{
        user: null,
        token: null,
        authLoading: false,
        requireLogin,
      }}
    >
      {ui}
    </LoginContext.Provider>
  )

export const renderAuthLoading = (ui: ReactElement) =>
  renderWithRouter(
    <LoginContext.Provider
      value={{ user: null, token: null, authLoading: true, requireLogin }}
    >
      {ui}
    </LoginContext.Provider>
  )
