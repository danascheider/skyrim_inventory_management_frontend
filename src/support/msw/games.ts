import { rest } from 'msw'
import { ResponseGame } from '../../types/apiData'
import { emptyGames, allGames } from '../data/games'

const BASE_URI = 'http://localhost:3000'

/**
 *
 * POST /games
 *
 */

export const postGamesSuccess = rest.post(
  `${BASE_URI}/games`,
  (req, res, ctx) => {
    const body = {
      id: 102,
      user_id: 412,
      name: req.params.name || 'My Game 3',
      description:
        req.params.description || 'This description is just for illustration.',
      created_at: new Date(),
      updated_at: new Date(),
    }

    return res(ctx.status(201), ctx.json(body))
  }
)

export const postGamesUnprocessable = rest.post(
  `${BASE_URI}/games`,
  (req, res, ctx) => {
    const body = {
      errors: [
        "Name can only contain alphanumeric characters, spaces, commas (,), hyphens (-), and apostrophes (')",
      ],
    }

    return res(ctx.status(422), ctx.json(body))
  }
)

export const postGamesServerError = rest.post(
  `${BASE_URI}/games`,
  (req, res, ctx) => {
    const body = { errors: ['oh noes'] }

    return res(ctx.status(500), ctx.json(body))
  }
)

/**
 *
 * GET /games
 *
 */

export const getGamesEmptySuccess = rest.get(
  `${BASE_URI}/games`,
  (req, res, ctx) => {
    return res(ctx.status(200), ctx.json(emptyGames))
  }
)

export const getGamesAllSuccess = rest.get(
  `${BASE_URI}/games`,
  (req, res, ctx) => {
    return res(ctx.status(200), ctx.json(allGames))
  }
)

export const getGamesServerError = rest.get(
  `${BASE_URI}/games`,
  (req, res, ctx) => {
    return res(ctx.status(500), ctx.json({ errors: ['Something went wrong'] }))
  }
)

/**
 *
 * PATCH /games/:id
 *
 */

const newOrExistingGame = (id: number): ResponseGame => {
  const existingGame = allGames.find((game) => game.id === id)

  if (existingGame) return existingGame

  return {
    id,
    user_id: 412,
    name: 'Skyrim Game',
    description: null,
    created_at: new Date(),
    updated_at: new Date()
  }
}

export const patchGameSuccess = rest.patch(
  `${BASE_URI}/games/:id`,
  async (req, res, ctx) => {
    const id = Number(req.params.id)
    const body = await req.json()
    const game = newOrExistingGame(id)

    return res(
      ctx.status(200),
      ctx.json({ ...game, ...body })
    )
  }
)

export const patchGameUnprocessableEntity = rest.patch(
  `${BASE_URI}/games/:id`,
  (req, res, ctx) => {
  return res(
    ctx.status(422),
    ctx.json({ errors: ['Name must be unique'] })
  )
})

export const patchGameNotFound = rest.patch(
  `${BASE_URI}/games/:id`,
  (req, res, ctx) => {
    return res(
      ctx.status(404)
    )
  }
)

export const patchGameServerError = rest.patch(
  `${BASE_URI}/games/:id`,
  (req, res, ctx) => {
    return res(
      ctx.status(500),
      ctx.json({ errors: ['oh noes'] })
    )
  }
)

/**
 *
 * DELETE /games/:id
 *
 */

export const deleteGameSuccess = rest.delete(
  `${BASE_URI}/games/:id`,
  (req, res, ctx) => {
    return res(ctx.status(204))
  }
)

export const deleteGameNotFound = rest.delete(
  `${BASE_URI}/games/:id`,
  (req, res, ctx) => {
    return res(ctx.status(404))
  }
)

export const deleteGameServerError = rest.delete(
  `${BASE_URI}/games/:id`,
  (req, res, ctx) => {
    return res(ctx.status(500), ctx.json({ errors: ['oh noes'] }))
  }
)
