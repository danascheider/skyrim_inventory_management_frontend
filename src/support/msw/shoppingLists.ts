import { rest } from 'msw'
import { allGames } from '../data/games'
import { allShoppingLists } from '../data/shoppingLists'
import { shoppingListsForGame } from '../data/shoppingLists'
import { newShoppingList, newShoppingListWithAggregate } from './helpers/data'

const BASE_URI = 'http://localhost:3000'
const gameIds = allGames.map(({ id }) => id)
const listIds = allShoppingLists.map(({ id }) => id)

/**
 *
 * POST /games/:game_id/shopping_lists
 *
 */

// Handles both 201 and 404 responses
export const postShoppingLists = rest.post(
  `${BASE_URI}/games/:gameId/shopping_lists`,
  async (req, res, ctx) => {
    const gameId: number = Number(req.params.gameId)

    if (gameIds.indexOf(gameId) < 0) return res(ctx.status(404))

    const attributes = await req.json()

    const responseBody = shoppingListsForGame(gameId).length
      ? newShoppingList(attributes, gameId)
      : newShoppingListWithAggregate(attributes, gameId)

    return res(ctx.status(201), ctx.json(responseBody))
  }
)

// Returns the same validation errors regardless of request body
// submitted
export const postShoppingListsUnprocessable = rest.post(
  `${BASE_URI}/games/:gameId/shopping_lists`,
  (_req, res, ctx) => {
    return res(
      ctx.status(422),
      ctx.json({
        errors: [
          'Title must be unique per game',
          "Title can only contain alphanumeric characters, spaces, commas (,), hyphens (-), and apostrophes (')",
        ],
      })
    )
  }
)

export const postShoppingListsServerError = rest.post(
  `${BASE_URI}/games/:gameId/shopping_lists`,
  (_req, res, ctx) => {
    return res(
      ctx.status(500),
      ctx.json({
        errors: ['Something went horribly wrong'],
      })
    )
  }
)

/**
 *
 * GET /games/:game_id/shopping_lists
 *
 */

// Covers both success and 404 cases
export const getShoppingLists = rest.get(
  `${BASE_URI}/games/:gameId/shopping_lists`,
  (req, res, ctx) => {
    const gameId: number = Number(req.params.gameId)
    if (gameIds.indexOf(gameId) < 0) return res(ctx.status(404))

    return res(
      ctx.status(200),
      ctx.json(shoppingListsForGame(Number(req.params.gameId)))
    )
  }
)

/**
 *
 * PATCH /shopping_lists/:id
 *
 */

// Covers both success and 404 cases
export const patchShoppingList = rest.patch(
  `${BASE_URI}/shopping_lists/:id`,
  async (req, res, ctx) => {
    const listId: number = Number(req.params.id)
    if (listIds.indexOf(listId) < 0) return res(ctx.status(404))

    const list = allShoppingLists.find(({ id }) => id === listId)
    const { title } = await req.json()

    return res(ctx.status(200), ctx.json({ ...list, title }))
  }
)

// Returns the same validation errors regardless of request
// body submitted
export const patchShoppingListUnprocessable = rest.patch(
  `${BASE_URI}/shopping_lists/:id`,
  (_req, res, ctx) => {
    return res(
      ctx.status(422),
      ctx.json({
        errors: [
          'Title must be unique per game',
          "Title can only contain alphanumeric characters, spaces, commas (,), hyphens (-), and apostrophes (')",
        ],
      })
    )
  }
)

export const patchShoppingListServerError = rest.patch(
  `${BASE_URI}/shopping_lists/:id`,
  (_req, res, ctx) => {
    return res(
      ctx.status(500),
      ctx.json({
        errors: ['Something went horribly wrong'],
      })
    )
  }
)

/**
 *
 * DELETE /shopping_lists/:id
 *
 */

// Covers both success and 404 cases
export const deleteShoppingList = rest.delete(
  `${BASE_URI}/shopping_lists/:listId`,
  (req, res, ctx) => {
    const listId = Number(req.params.listId)
    const list = allShoppingLists.find(({ id }) => id === listId)

    if (!list) return res(ctx.status(404))

    const lists = shoppingListsForGame(list.game_id)

    let json
    if (lists.length === 2) {
      json = {
        deleted: lists.map(({ id }) => id),
      }
    } else {
      json = {
        deleted: [list.id],
        aggregate: lists[0],
      }
    }

    return res(ctx.status(200), ctx.json(json))
  }
)

export const deleteShoppingListServerError = rest.delete(
  `${BASE_URI}/shopping_lists/:listId`,
  (_req, res, ctx) => {
    return res(
      ctx.status(500),
      ctx.json({
        errors: ['Something went horribly wrong'],
      })
    )
  }
)
