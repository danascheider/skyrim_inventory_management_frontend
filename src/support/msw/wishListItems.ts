import { http } from 'msw'
import { allWishLists, wishListsForGame } from '../data/wishLists'
import { allWishListItems } from '../data/wishListItems'
import { newWishListItem } from './helpers/data'
import { type RequestWishListItem } from '../../types/apiData'

const BASE_URI = 'http://localhost:3000'
const listIds = allWishLists.map(({ id }) => id)

/**
 *
 * POST /wish_lists/:list_id/list_items
 *
 */

// Handles 201 and 404 responses
export const postWishListItemsSuccess = http.post(
  `${BASE_URI}/wish_lists/:listId/wish_list_items`,
  async ({ request, params }) => {
    const listId = Number(params.listId)

    if (listIds.indexOf(listId) < 0) return new Response(null, { status: 404 })

    const attributes = await request.json() as RequestWishListItem

    return new Response(
      JSON.stringify(newWishListItem(attributes, listId)),
      { status: 201 }
    )
  }
)

// Returns the same validation errors regardless of request body
// submitted
export const postWishListItemsUnprocessable = http.post(
  `${BASE_URI}/wish_lists/:listId/wish_list_items`,
  (_) => {
    return new Response(
      JSON.stringify({
        errors: [
          'Quantity must be greater than 0',
          'Unit weight must be greater than or equal to 0',
        ],
      }),
      { status: 422 }
    )
  }
)

export const postWishListItemsServerError = http.post(
  `${BASE_URI}/wish_lists/:listId/wish_list_items`,
  (_) => {
    return new Response(
      JSON.stringify({
        errors: ['Something went horribly wrong'],
      }),
      { status: 500 }
    )
  }
)

/**
 *
 * PATCH /wish_list_items/:id
 *
 */

export const incrementWishListItemSuccess = http.patch(
  `${BASE_URI}/wish_list_items/:id`,
  ({ params }) => {
    const itemId: number = Number(params.id)
    const item = allWishListItems.find(({ id }) => id === itemId)
    const list = allWishLists.find(({ id }) => id === item?.list_id)

    if (!item || !list) return new Response(null, { status: 404 })

    const allItems = wishListsForGame(list.game_id).flatMap(
      ({ list_items }) =>
        list_items.filter(({ description }) => description === item.description)
    )

    const aggListItem = { ...allItems[0], quantity: allItems[0].quantity + 1 }
    const regItem = { ...item, quantity: item.quantity + 1 }

    return new Response(JSON.stringify([aggListItem, regItem]), { status: 200 })
  }
)

export const decrementWishListItemSuccess = http.patch(
  `${BASE_URI}/wish_list_items/:id`,
  ({ request, params }) => {
    const itemId: number = Number(params.id)
    const item = allWishListItems.find(({ id }) => id === itemId)
    const list = allWishLists.find(({ id }) => id === item?.list_id)

    if (!item || !list) return new Response(null, { status: 404 })

    const allItems = wishListsForGame(list.game_id).flatMap(
      ({ list_items }) =>
        list_items.filter(({ description }) => description === item.description)
    )
    const aggListItem = { ...allItems[0], quantity: allItems[0].quantity - 1 }
    const regItem = { ...item, quantity: item.quantity - 1 }

    return new Response(JSON.stringify([aggListItem, regItem]), { status: 200 })
  }
)

// Logic for updating wish list items is too complex to model in an
// MSW handler. This handler is written under the assumption that only
// the list item's "notes" have been updated in the request. Updates to
// the other values are not supported by this handler. In particular,
// updates to unit weight are complex and should not be used in front-end
// tests if it can be avoided.
//
// This handler further assumes that the aggregate list item associated with
// the list item being updated has no "notes" value other than that of the
// list item being updated - i.e., none of its other associated list items, if
// any, have notes. If they did, that, too, would further complicate things.
export const updateWishListItemSuccess = http.patch(
  `${BASE_URI}/wish_list_items/:id`,
  async ({ request, params }) => {
    const itemId: number = Number(params.id)
    const item = allWishListItems.find(({ id }) => id === itemId)
    const list = allWishLists.find(({ id }) => id === item?.list_id)

    if (!item || !list) return new Response(null, { status: 404 })

    const json = await request.json() as RequestWishListItem

    const allItems = wishListsForGame(list.game_id).flatMap(
      ({ list_items }) =>
        list_items.filter(({ description }) => description === item.description)
    )
    const aggListItem = { ...allItems[0], notes: json.notes }
    const regItem = { ...item, notes: json.notes }

    return new Response(JSON.stringify([aggListItem, regItem]), { status: 200 })
  }
)

// Returns the same validation errors regardless of request body
export const updateWishListItemUnprocessable = http.patch(
  `${BASE_URI}/wish_list_items/:id`,
  (_) => {
    return new Response(
      JSON.stringify({
        errors: [
          'Quantity must be greater than 0',
          'Unit weight must be greater than or equal to 0',
        ],
      }),
      { status: 422 }
    )
  }
)

export const updateWishListItemServerError = http.patch(
  `${BASE_URI}/wish_list_items/:id`,
  (_) => {
    return new Response(
      JSON.stringify({
        errors: ['Something went horribly wrong'],
      }),
      { status: 500 }
    )
  }
)

/**
 *
 * DELETE /wish_list_items/:id
 *
 */

// Note: If the list item, list, and aggregate list are all
//       found in the test data, this will delete the item
//       from the regular list AND the aggregate list regardless
//       of whether there are other matching items. This function
//       does not match the complexity of back-end behaviour.
export const deleteWishListItemSuccess = http.delete(
  `${BASE_URI}/wish_list_items/:id`,
  ({ request, params }) => {
    const itemId = Number(params.id)
    const item = allWishListItems.find(({ id }) => id === itemId)

    if (!item) return new Response(null, { status: 404 })

    const listId = item.list_id
    const regList = allWishLists.find(({ id }) => id === listId)

    if (!regList) return new Response(null, { status: 404 })

    const aggListId = regList.aggregate_list_id
    const aggList = allWishLists.find(({ id }) => id === aggListId)

    if (!aggList) return new Response(null, { status: 404 })

    const list = { ...regList }
    const aggregate = { ...aggList }

    list.list_items = list.list_items.filter(({ id }) => id !== itemId)
    aggregate.list_items = aggregate.list_items.filter(
      ({ description }) => description !== item.description
    )

    return new Response(JSON.stringify([aggregate, list]), { status: 200 })
  }
)

export const deleteWishListItemServerError = http.delete(
  `${BASE_URI}/wish_list_items/:id`,
  (_) => {
    return new Response(
      JSON.stringify({ errors: ['Something went horribly wrong'] }),
      { status: 500 }
    )
  }
)
