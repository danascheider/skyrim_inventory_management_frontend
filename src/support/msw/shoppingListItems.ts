import { rest } from 'msw'
import { allShoppingLists, shoppingListsForGame } from '../data/shoppingLists'
import { allShoppingListItems } from '../data/shoppingListItems'
import { newShoppingListItem } from './helpers/data'
import { RequestShoppingListItem } from '../../types/apiData'

const BASE_URI = 'http://localhost:3000'
const listIds = allShoppingLists.map(({ id }) => id)

/**
 *
 * POST /shopping_lists/:list_id/list_items
 *
 */

// Handles 201 and 404 responses
export const postShoppingListItemsSuccess = rest.post(
  `${BASE_URI}/shopping_lists/:listId/shopping_list_items`,
  async (req, res, ctx) => {
    const listId: number = Number(req.params.listId)

    if (listIds.indexOf(listId) < 0) return res(ctx.status(404))

    const attributes = await req.json()

    return res(
      ctx.status(201),
      ctx.json(newShoppingListItem(attributes, listId))
    )
  }
)

// Returns the same validation errors regardless of request body
// submitted
export const postShoppingListItemsUnprocessable = rest.post(
  `${BASE_URI}/shopping_lists/:listId/shopping_list_items`,
  (_req, res, ctx) => {
    return res(
      ctx.status(422),
      ctx.json({
        errors: [
          'Quantity must be greater than 0',
          'Unit weight must be greater than or equal to 0',
        ],
      })
    )
  }
)

export const postShoppingListItemsServerError = rest.post(
  `${BASE_URI}/shopping_lists/:listId/shopping_list_items`,
  (_req, res, ctx) => {
    return res(
      ctx.status(500),
      ctx.json({
        errors: ['Something went horribly wrong'],
      })
    )
  }
)

/**
 *
 * PATCH /shopping_list_items/:id
 *
 */

export const incrementShoppingListItemSuccess = rest.patch(
  `${BASE_URI}/shopping_list_items/:id`,
  (req, res, ctx) => {
    const itemId: number = Number(req.params.id)
    const item = allShoppingListItems.find(({ id }) => id === itemId)
    const list = allShoppingLists.find(({ id }) => id === item?.list_id)

    if (!item || !list) return res(ctx.status(404))

    const allItems = shoppingListsForGame(list.game_id).flatMap(
      ({ list_items }) =>
        list_items.filter(({ description }) => description === item.description)
    )

    const aggListItem = { ...allItems[0], quantity: allItems[0].quantity + 1 }
    const regItem = { ...item, quantity: item.quantity + 1 }

    return res(ctx.status(200), ctx.json([aggListItem, regItem]))
  }
)

export const decrementShoppingListItemSuccess = rest.patch(
  `${BASE_URI}/shopping_list_items/:id`,
  (req, res, ctx) => {
    const itemId: number = Number(req.params.id)
    const item = allShoppingListItems.find(({ id }) => id === itemId)
    const list = allShoppingLists.find(({ id }) => id === item?.list_id)

    if (!item || !list) return res(ctx.status(404))

    const allItems = shoppingListsForGame(list.game_id).flatMap(
      ({ list_items }) =>
        list_items.filter(({ description }) => description === item.description)
    )
    const aggListItem = { ...allItems[0], quantity: allItems[0].quantity - 1 }
    const regItem = { ...item, quantity: item.quantity - 1 }

    return res(ctx.status(200), ctx.json([aggListItem, regItem]))
  }
)

// Logic for updating shopping list items is too complex to model in an
// MSW handler. This handler is written under the assumption that only
// the list item's "notes" have been updated in the request. Updates to
// the other values are not supported by this handler. In particular,
// updates to unit weight are complex and should not be used in front-end
// tests if it can be avoided.
//
// This handler further assumes that the aggregate list item associated with
// the list item being updated has no "notes" value other than that of the
// list item being updated - i.e., none of its other associated list items, if
// any, have notes. If they did, that, too, would further complicate things.
export const updateShoppingListItemSuccess = rest.patch(
  `${BASE_URI}/shopping_list_items/:id`,
  async (req, res, ctx) => {
    const itemId: number = Number(req.params.id)
    const item = allShoppingListItems.find(({ id }) => id === itemId)
    const list = allShoppingLists.find(({ id }) => id === item?.list_id)

    if (!item || !list) return res(ctx.status(404))

    const json = (await req.json()) as RequestShoppingListItem

    const allItems = shoppingListsForGame(list.game_id).flatMap(
      ({ list_items }) =>
        list_items.filter(({ description }) => description === item.description)
    )
    const aggListItem = { ...allItems[0], notes: json.notes }
    const regItem = { ...item, notes: json.notes }

    return res(ctx.status(200), ctx.json([aggListItem, regItem]))
  }
)

// Returns the same validation errors regardless of request body
export const updateShoppingListItemUnprocessable = rest.patch(
  `${BASE_URI}/shopping_list_items/:id`,
  (_req, res, ctx) => {
    return res(
      ctx.status(422),
      ctx.json({
        errors: [
          'Quantity must be greater than 0',
          'Unit weight must be greater than or equal to 0',
        ],
      })
    )
  }
)

export const updateShoppingListItemServerError = rest.patch(
  `${BASE_URI}/shopping_list_items/:id`,
  (_req, res, ctx) => {
    return res(
      ctx.status(500),
      ctx.json({
        errors: ['Something went horribly wrong'],
      })
    )
  }
)

/**
 *
 * DELETE /shopping_list_items/:id
 *
 */

// Note: If the list item, list, and aggregate list are all
//       found in the test data, this will delete the item
//       from the regular list AND the aggregate list regardless
//       of whether there are other matching items. This function
//       does not match the complexity of back-end behaviour.
export const deleteShoppingListItemSuccess = rest.delete(
  `${BASE_URI}/shopping_list_items/:id`,
  (req, res, ctx) => {
    const itemId: number = Number(req.params.id)
    const item = allShoppingListItems.find(({ id }) => id === itemId)

    if (!item) return res(ctx.status(404))

    const listId = item.list_id
    const regList = allShoppingLists.find(({ id }) => id === listId)

    if (!regList) return res(ctx.status(404))

    const aggListId = regList.aggregate_list_id
    const aggList = allShoppingLists.find(({ id }) => id === aggListId)

    if (!aggList) return res(ctx.status(404))

    const list = { ...regList }
    const aggregate = { ...aggList }

    list.list_items = list.list_items.filter(({ id }) => id !== itemId)
    aggregate.list_items = aggregate.list_items.filter(
      ({ description }) => description !== item.description
    )

    return res(ctx.status(200), ctx.json([aggregate, list]))
  }
)

export const deleteShoppingListItemServerError = rest.delete(
  `${BASE_URI}/shopping_list_items/:id`,
  (_req, res, ctx) => {
    return res(
      ctx.status(500),
      ctx.json({ errors: ['Something went horribly wrong'] })
    )
  }
)
