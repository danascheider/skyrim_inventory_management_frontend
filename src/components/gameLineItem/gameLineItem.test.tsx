import { describe, test, expect, vitest, afterEach, beforeEach } from 'vitest'
import { render, act } from '@testing-library/react'
import { setupServer } from 'msw/node'
import { deleteGameSuccess, getGamesAllSuccess } from '../../support/msw/games'
import {
  GamesContext,
  type GamesContextType,
} from '../../contexts/gamesContext'
import GameLineItem from './gameLineItem'
import { renderAuthenticated } from '../../support/testUtils'

const contextValue: GamesContextType = {
  games: [
    {
      id: 4,
      user_id: 42,
      name: 'De finibus bonorum et malorum',
      description: 'This is my game',
      created_at: new Date(),
      updated_at: new Date(),
    },
  ],
  gamesLoadingState: 'DONE',
  createGame: () => {},
  destroyGame: () => {},
}

// This component can't really be tested because Vitest fucking sucks.
// It won't get the style off the element, even if the style is computed
// with JavaScript, and because react-animate-height uses a CSS-based
// approach rather than adding and removing an element, there's just no
// way to test whether the user can see the description or not.
describe('GameLineItem', () => {
  test('matches snapshot', () => {
    const wrapper = render(
      <GamesContext.Provider value={contextValue}>
        <GameLineItem
          gameId={4}
          name="De finibus bonorum et malorum"
          description="This is my game"
        />
      </GamesContext.Provider>
    )
    expect(wrapper).toBeTruthy()
    expect(wrapper).toMatchSnapshot()
  })

  describe('destroying the game', () => {
    beforeEach(() => {
      contextValue.destroyGame = vitest
        .fn()
        .mockImplementation((_gameId: number) => {})
    })

    test('destroys the game when the button is clicked', () => {
      const wrapper = renderAuthenticated(
        <GamesContext.Provider value={contextValue}>
          <GameLineItem
            gameId={4}
            name="De finibus bonorum et malorum"
            description="This is my game"
          />
        </GamesContext.Provider>
      )

      window.confirm = vitest.fn().mockImplementation(() => true)

      const xIcon = wrapper.getByTestId('destroyGame4')
      act(() => xIcon.click())

      expect(contextValue.destroyGame).toHaveBeenCalledWith(4)
    })

    test("doesn't destroy the game when the user cancels", () => {
      const wrapper = renderAuthenticated(
        <GamesContext.Provider value={contextValue}>
          <GameLineItem
            gameId={4}
            name="De finibus bonorum et malorum"
            description="This is my game"
          />
        </GamesContext.Provider>
      )

      window.confirm = vitest.fn().mockImplementation(() => false)

      const xIcon = wrapper.getByTestId('destroyGame4')
      act(() => xIcon.click())

      expect(contextValue.destroyGame).not.toHaveBeenCalled()
    })
  })
})
