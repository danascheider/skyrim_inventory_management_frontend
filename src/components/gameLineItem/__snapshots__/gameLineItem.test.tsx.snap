// Vitest Snapshot v1

exports[`GameLineItem > matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_root_fd0dcf"
      >
        <div
          class="_summary_fd0dcf"
        >
          <button
            class="_icon_fd0dcf"
            data-testid="destroyGame4"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-xmark _fa_fd0dcf"
              data-icon="xmark"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 320 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M310.6 150.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L160 210.7 54.6 105.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L114.7 256 9.4 361.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L160 301.3 265.4 406.6c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L205.3 256 310.6 150.6z"
                fill="currentColor"
              />
            </svg>
          </button>
          <h3
            aria-controls="game4Details"
            aria-expanded="false"
            class="_header_fd0dcf"
            role="button"
          >
            De finibus bonorum et malorum
          </h3>
        </div>
        <div
          aria-hidden="true"
          class="rah-static rah-static--height-zero "
          id="game4Details"
          style="height: 0; overflow: hidden;"
        >
          <div
            style="display: none;"
          >
            <p
              class="_description_fd0dcf"
            >
              This is my game
            </p>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="_root_fd0dcf"
    >
      <div
        class="_summary_fd0dcf"
      >
        <button
          class="_icon_fd0dcf"
          data-testid="destroyGame4"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-xmark _fa_fd0dcf"
            data-icon="xmark"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 320 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M310.6 150.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L160 210.7 54.6 105.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L114.7 256 9.4 361.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L160 301.3 265.4 406.6c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L205.3 256 310.6 150.6z"
              fill="currentColor"
            />
          </svg>
        </button>
        <h3
          aria-controls="game4Details"
          aria-expanded="false"
          class="_header_fd0dcf"
          role="button"
        >
          De finibus bonorum et malorum
        </h3>
      </div>
      <div
        aria-hidden="true"
        class="rah-static rah-static--height-zero "
        id="game4Details"
        style="height: 0; overflow: hidden;"
      >
        <div
          style="display: none;"
        >
          <p
            class="_description_fd0dcf"
          >
            This is my game
          </p>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
